### PR TITLE
add email_login tool and add email summarization scenario example

### PR DESCRIPTION
**Pull request by FromCSUZhou**
**Features**

- add email_login tool for code_interpreter
- add email summarization scenario example
    
**Feature Docs**

**Influence**

**Result**

**Other**

### CODE REVIEW

File: examples/email_summary.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+@Author  :   Tuo Zhou
+@File    :   email_summary.py
+"""
+
+from metagpt.roles.ci.code_interpreter import CodeInterpreter
+
+
+async def main():
+    # prompt_response = """I will give you your Outlook email account(englishgpt@outlook.com) and password(the outlook_email_password item in the environment variable). You need to find the latest email in my inbox with the sender's suffix @qq.com and reply to him "Thank you! I have received your email~"""""
+    prompt_summary = """I will give you your Outlook email account(englishgpt@outlook.com) and password(outlook_email_password item in the environment variable).
```
Comment (garylin2099) [inline examples/email_summary.py:13]: 应该叫prompt，非prompt_summary。去除不用的注释

File: examples/email_summary.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+            Firstly, Please help me present the latest 5 senders and full letter contents.
+            Then, summarize each of the 5 emails into one sentence with Chinese(you can do this by yourself, don't need import other models to do this) and output them in a markdown format."""
+    # ci_response = CodeInterpreter(goal=prompt_response, use_tools=True)
+    ci_summary = CodeInterpreter(goal=prompt_summary, use_tools=True)
```
Comment (garylin2099) [inline examples/email_summary.py:17]: 这里不需要给定goal了

File: metagpt/tools/libs/email_login.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    Args:
+        email_address (str): Email address that needs to be logged in and linked.
+        email_password (str): Password for the email address that needs to be logged in and linked.
+
+    Returns:
+        object: The imap_tools's MailBox object returned after successfully connecting to the mailbox through imap_tools, including various information about this account (email, etc.), or None if login fails.
+    """
+
+    # Define a dictionary mapping email domains to their IMAP server addresses
+    imap_servers = {
```
Comment (garylin2099) [inline metagpt/tools/libs/email_login.py:21]: 常量的话，放在函数外面用IMAP_SERVERS命名

File: metagpt/tools/libs/email_login.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    # Extract the domain from the email address
+    domain = email_address.split("@")[-1]
+
+    # Determine the correct IMAP server
+    imap_server = imap_servers.get(domain)
+
+    if not imap_server:
+        print(f"IMAP server for {domain} not found.")
```
Comment (garylin2099) [inline metagpt/tools/libs/email_login.py:47]: from metagpt.logs import logger. 用logger.warning, logger.error, logger.info

**Comment by garylin2099** [comment]
加单测测email_login这个py

File: tests/metagpt/tools/libs/test_email_login.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+correct_email_address = "englishgpt@outlook.com"
+correct_email_password = os.environ.get("outlook_email_password")
+incorrect_email_address = "test@unknown.com"
+incorrect_email_password = "incorrect_password"
+
+
+@pytest.fixture
+def imap_server_setup():
+    # Use patch to mock the behavior of MailBox from the correct module path
+    with patch("metagpt.tools.libs.email_login.MailBox") as mock_mailbox:
```
Comment (garylin2099) [inline tests/metagpt/tools/libs/test_email_login.py:21]: 用pytest的mocker

File: examples/email_summary.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+@Date    :   2024/02/07 
+@Author  :   Tuo Zhou
+@File    :   email_summary.py
+"""
+
+from metagpt.roles.ci.code_interpreter import CodeInterpreter
+
+
+async def main():
+    # For email response prompt
```
Comment (garylin2099) [inline examples/email_summary.py:12]: - 用一个变量email_account = "your_email_account"，prompt内引用。这是提醒用户换自己的邮箱
- outlook_email_password -> email_password
- with chinese去掉
- don't need -> no need

**Comment by garylin2099** [comment]
贴下测试成功的截图

File: metagpt/tools/libs/email_login.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        logger.error(f"IMAP server for {domain} not found.")
+        return None
+
+    # Attempt to log in to the email account
+    try:
+        mailbox = MailBox(imap_server).login(email_address, email_password)
+        print("Login successful")
+        return mailbox
+    except Exception as e:
+        print(f"Login failed: {e}")
```
Comment (garylin2099) [inline metagpt/tools/libs/email_login.py:57]: logger

**Comment by FromCSUZhou** [comment]
image

**Pull request merged**